### PR TITLE
blog.md uploaded

### PR DESCRIPTION
**Pull request by xuanzhaogao**

### CODE REVIEW

File: blog.md [hunk 1]
```diff
@@ -0,0 +1,203 @@
+# CuTropicalGEMM.jl: A fast package for Matrix Multiplication of Tropical Numbers on GPU
+
+In this blog, I would like to share some experience of developing a julia package called [CuTropicalGemm.jl](github.com/TensorBFS/CuTropicalGEMM.jl), which is used to calculate Generic Matrix Multiplication (GEMM) of Tropical Numbers on Nvidia GPUs.
```
Comment (GiggleLiu) [inline blog.md]: It is better to start with a promise. "you will learn..." followed by a table of content.
1. What is Tropical algebra and Tropical matrix multplication and why we need them
2. How to implement a fast generic matrix multiplication on Nvidia GPU
3. ...

**Important:** Please credit the OSPP project! e.g.
This blog is a technical note for the [Open Source Promotion Plan 2023](https://summer-ospp.ac.cn/) project ["TropicalGEMM on GPU"](https://summer-ospp.ac.cn/org/prodetail/23fec0105?lang=en&list=pro) that relesed by JuliaCN, where I developed a [Julia](https://julialang.org/) package [CuTropicalGemm.jl](github.com/TensorBFS/CuTropicalGEMM.jl) calculate Generic Matrix Multiplication (GEMM) of Tropical Numbers on Nvidia GPUs.

File: blog.md [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+In the following sections, I will introduce the background of this package, the technique we used and show some examples.
+
+## What are tropical numbers and why we need them?
+
+Tropical numbers, also known as the tropical semiring, is defined on $\mathbb{R} \cup \{\infty\}$ with addition and multiplication defined as $x \oplus y = \max\{(x,~y)\},~x \otimes y = x + y$, this algebra is also called the TropicalMaxPlus.
```
Comment (GiggleLiu) [inline blog.md]: Explain the meaning of "semi". Unlike a regular ring algebra that we familiar with, the tropical plus operation (max) does not have an inverse.

File: blog.md [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+The following properties holds the tropical numbers
+
+1. Addition $\oplus$ is commutative monoid and $0$ is its neutral element of addition.
+2. Multiplication $\otimes$ is monoid and $0$ is its neutral element of multiplication.
+3. Multiplication left and right distributes over addition.
+
+In recent years, the tropical numbers have been widely used in various areas, including optimization, physics, and computer science, due to its computational simplicity.
+
+For example, it was shown that solving the groud state energy of spin glass problem can be mapped as contraction of a tropical tensor network, which is actually calculating tons of matrix multiplication.
```
Comment (GiggleLiu) [inline blog.md]: We should include some machine learning applications too!

File: blog.md [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+Although such reformulation did not actucally reduce the complexity of the problem, but in this way we can fully use the power of parallel computing technology developed in recent years and greatly speed up the computation.
+
+For that purpose, a fast implementation of the TropicalGEMM on GPU is on demand.
```
Comment (GiggleLiu) [inline blog.md]: This blog may have two purposes
1. attract potential researchers to use `CuTropicalGEMM`.
2. technical note that may inspire other developers and contributers.

I wish we can serve for the first purpose first, we should show how to use it and how to check its performance. Then we can focus on why it can achieve such a good performance.

File: blog.md [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+For example, when handling a matrix multiplication with size $4 \times 4 \times 10^6$, what is actually calculated is $64 \times 32 \times 10^6$, which means that the code will waste a lot of time on the padding elements.
+Unfortunately, such narrow matrices are very common in the tensor network contraction process, and the performance of the code on narrow matrices is very important.
+
+Now we are considering further optimize the code for narrow matrices, related code are stored in the branch [narrow matrices](https://github.com/TensorBFS/CuTropicalGEMM.jl/tree/narrow_matrices).
+
+## Acknowledgement
+
+I am very grateful for the guidance and assistance provided by Professor Liu during the project implementation process.
+I would like to thank Tim Besard for his invaluable guidance and support during the development of the package, his expertise in GPU utilization have been immensely helpful. 
+I also want to thank Tyler Thomas for his assistance in understanding the usage of BinaryBuilder.jl.
```
Comment (GiggleLiu) [inline blog.md]: Add links to their github profiles.

File: blog.md [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+Although the tropical algebra can not use the fused multiple add core (FMA), we found that the operation add/mul and max/min can be done on FMA and ALU parallelly, which means that we can use the FMA to do the add/mul and ALU to do the max/min at the same time.
+Then after all the calculation is done, the target in the registers will be stored back to global memory directly.
+
+For the boundary elements, a padding strategy is used, we simply set the element which are not acctually in the matrix as the zero element of the corresponding algebra, so that they will not effect the result of the calculation.
+In our package, we set the parameters as
+$$
+    BM = 64,~BK = 32,~BN = 64,~TM = TN = 4.
+$$
+
+In the end, we wrapped the code in Julia, and overloaded the function `LinearAlgebra.mutmul!`, so that a simple $*$ for tropical matrices will call our function and use GPU in calculations.
```
Comment (GiggleLiu) [inline blog.md]:
```suggestion
In the end, we wrapped the code in Julia, and overloaded the function `LinearAlgebra.mul!`, so that a simple $*$ for tropical matrices will call our function and use GPU in calculations.
```

File: blog.md (hunk 1 above)
Comment (GiggleLiu) [reply blog.md]: It is also good to add a one line introduction to Julia programming language, because most people are not familiar with it.

File: blog.md [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+2-element Vector{Tropical{Float64}}:
+ 0.5502018201390529ₜ
+ 0.5321062079648163ₜ
+```
+Since we define the tropical number as a subtype of `Number`,
+```julia
+julia> isbitstype(Tropical{Float64})
+true
+```
+which means the storage of tropical number in the memory is continium.
```
Comment (GiggleLiu) [inline blog.md]:
```suggestion
which means the storage of an array of tropical numbers in the memory is continium.
```

File: blog.md [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+struct Tropical{T} <: AbstractSemiring
+    n::T
+
+    Tropical{T}(x) where T = new{T}(T(x))
+    function Tropical(x::T) where T
+        new{T}(x)
+    end
+end
+```
+and then we overloaded the operations
```
Comment (GiggleLiu) [inline blog.md]: Missing keyword: multiple dispatch

File: blog.md [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+true
+```
+which means the storage of tropical number in the memory is continium.
+
+All these things work naturely, users will be able to use the tropical algebra just like real numbers.
+
+### High performance CUDA kernel
+
+For the lower level kernel, due to the requirement of performance, we choose to use `C-Cuda` directly and wrap the code with a Julia interface.
+In this part, we simply used some skills common in GEMM kernels to speed up the code, and we learned a lot from the repo [CUDA_gemm](https://github.com/Cjkkkk/CUDA_gemm) by [Cjkkk](https://cjkkkk.github.io).
```
Comment (GiggleLiu) [inline blog.md]:
```suggestion
In this part, we simply used some techniques common in GEMM kernels to speed up the code, and we learned a lot from the repo [CUDA_gemm](https://github.com/Cjkkkk/CUDA_gemm) by [Cjkkk](https://cjkkkk.github.io).
```

File: blog.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+Here we will briefly introduce these skills, and for more detailed introduction, I recommand this [blog](https://zhuanlan.zhihu.com/p/441146275) in Chinese.
+
+As we all know, GPU is fast because it have a large amount of cores, and the cores are grouped into blocks.
```
Comment (GiggleLiu) [inline blog.md]:
```suggestion
GPU is fast because it have a large amount of CUDA cores, and the cores are grouped into blocks.
```

Comment (GiggleLiu) [reply blog.md]: "CUDA core" is not "core".

File: blog.md [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+For the boundary elements, a padding strategy is used, we simply set the element which are not acctually in the matrix as the zero element of the corresponding algebra, so that they will not effect the result of the calculation.
+In our package, we set the parameters as
+$$
+    BM = 64,~BK = 32,~BN = 64,~TM = TN = 4.
+$$
+
+In the end, we wrapped the code in Julia, and overloaded the function `LinearAlgebra.mutmul!`, so that a simple $*$ for tropical matrices will call our function and use GPU in calculations.
+Here is an example:
+```julia
+julia> using CUDA, LinearAlgebra, TropicalNumbers, CuTropicalGEMM
```
Comment (GiggleLiu) [inline blog.md:160]: Maybe we should re-export the types in TropicalNumbers to CuTropicalGEMM, so that we can have one less package to import.

File: blog.md (hunk 10 above)
Comment (GiggleLiu) [reply blog.md]: "large amount of" is a bit vague. Please use some examples, e.g. A800 has xxx CUDA cores, the peak flops is xxx. Comparing with CPU, ...

File: blog.md [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+### High performance CUDA kernel
+
+For the lower level kernel, due to the requirement of performance, we choose to use `C-Cuda` directly and wrap the code with a Julia interface.
+In this part, we simply used some skills common in GEMM kernels to speed up the code, and we learned a lot from the repo [CUDA_gemm](https://github.com/Cjkkkk/CUDA_gemm) by [Cjkkk](https://cjkkkk.github.io).
+Here we will briefly introduce these skills, and for more detailed introduction, I recommand this [blog](https://zhuanlan.zhihu.com/p/441146275) in Chinese.
+
+As we all know, GPU is fast because it have a large amount of cores, and the cores are grouped into blocks.
+In the GPU kernels, we can allocate tasks to different blocks by decomposing them into a series of unrelated sub-tasks. Furthermore, we can further decompose the sub-tasks and assign them to threads for execution.
+However, this large number of parallel tasks also leads to another issue, which is the significant data payload. 
+Although the shared memory accessible by blocks is similar in speed to the L2 cache in CPUs, its capacity is limited.
```
Comment (GiggleLiu) [inline blog.md]: You did not introduced shared memory.

File: blog.md [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+
+As we all know, GPU is fast because it have a large amount of cores, and the cores are grouped into blocks.
+In the GPU kernels, we can allocate tasks to different blocks by decomposing them into a series of unrelated sub-tasks. Furthermore, we can further decompose the sub-tasks and assign them to threads for execution.
+However, this large number of parallel tasks also leads to another issue, which is the significant data payload. 
+Although the shared memory accessible by blocks is similar in speed to the L2 cache in CPUs, its capacity is limited. 
+Therefore, it is not possible to load all data into shared memory at once, and only the data required for the current block calculation can be loaded each time.
+Unfortunately, GEMM is a very memory-intensive operation, for example, when calculating the GEMM between a $M \times K$ matrix and a $K \times N$ matrix, if we use the naive way, i.e. evaluate the element in the result matrix one by one, we will have to load $M \times K \times N$ elements from the slow global memeory directly to registers in the whole process, and this generally far exceeds the data bandwidth of the GPU, resulting in severe performance issues.
```
Comment (GiggleLiu) [inline blog.md]: There is a gap between "evaluate the element in the result matrix one by one" and "we will have to load $M \times K \times N$ elements". You'd better add one more sentence: the data payload for each element is K, and the total payload is...

File: blog.md [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+
+To avoid the heavy data loading, we first split the target matrix into blocks with size $BM \times BN$, and each GPU block will be used to calculate one of the tiled matrix, as shown in the fig below:
```
Comment (GiggleLiu) [inline blog.md]: missing key word: tiling

File: blog.md [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+
+Then we replaced the operations directly.
+As we mentioned above, the `Tropical` type is `bitstype`, and the data storaged in memeories are simply the floating point numbers, which can be directly used by CUDA kernels.
+Although the tropical algebra can not use the fused multiple add core (FMA), we found that the operation add/mul and max/min can be done on FMA and ALU parallelly, which means that we can use the FMA to do the add/mul and ALU to do the max/min at the same time.
+Then after all the calculation is done, the target in the registers will be stored back to global memory directly.
+
+For the boundary elements, a padding strategy is used, we simply set the element which are not acctually in the matrix as the zero element of the corresponding algebra, so that they will not effect the result of the calculation.
+In our package, we set the parameters as
+$$
+    BM = 64,~BK = 32,~BN = 64,~TM = TN = 4.
```
Comment (GiggleLiu) [inline blog.md]: Maybe move this equation to the above?

File: blog.md [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+julia> using TropicalGEMM
+
+julia> @time A * B;
+  0.095628 seconds (2 allocations: 8.000 MiB)
+
+julia> CuA = CuArray(A);
+
+julia> CuB = CuArray(B);
+
+julia> @time CuA * CuB;
```
Comment (GiggleLiu) [inline blog.md]: 1. Please emphasis a second run.
2. It is not enough for compare wtih CPU, please also compare with the fall back implementation with map-reduce.
3. I remember that you have removed the synchronize operation? Why can you use `@time` here?

File: blog.md [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+
+### Benchmark
+
+A simple benchmark is conducted to test the performance of our package, as shown below.
+We compared the performance of `CuTropicalGEMM.jl`, [GemmKernels.jl](https://github.com/JuliaGPU/GemmKernels.jl) and direct CUDA.jl map reduce on Tropical GEMM with single precision.
+The test is done on using NVIDIA A800 80GB PCIe, and the performance of Cublas on normal GEMM is used as a reference of the maximum computing power.
+
+![Benchmark](figs/benchmark.png)
+
+Further benchmarking is still in development, and will be uoloaded to this repo [CuTropicalGEMM_benchmark](https://github.com/ArrogantGao/CuTropicalGEMM_benchmark).
```
Comment (GiggleLiu) [inline blog.md:296]:
```suggestion
Further benchmarking is still in development, and will be uploaded to this repo [CuTropicalGEMM_benchmark](https://github.com/ArrogantGao/CuTropicalGEMM_benchmark).
```

File: blog.md [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+
+## Further Optimization
```
Comment (GiggleLiu) [inline blog.md]:
```suggestion
## Optimization of narrow matrix performance
```

File: blog.md [hunk 19]
```diff
@@ -0,0 +1,203 @@
+# CuTropicalGEMM.jl: A fast package for Matrix Multiplication of Tropical Numbers on GPU
```
Comment (GiggleLiu) [inline blog.md]:
```suggestion
# CuTropicalGEMM.jl: The fasted package for Tropical Matrix Multiplication on GPU
```

Comment (GiggleLiu) [reply blog.md]: Or we can have a more sexy title:
"How to implement generic matrix multiplication (GEMM) with generic element types on GPU"

**Review by GiggleLiu — COMMENTED** [review]
I really enjoyed reading this blog post! In general, I wish it could include more links to external resouces. Such as the machine learning papers related to Tropical numbers, Julia packages. For tutorial purpose, I think it is better to include some description about how to eastimate the peak flops of GPU, and how do we measure the performance in practise to help readers appreciate this work better.

**Comment by GiggleLiu** [comment]
Oh, you forgot to list the table of tropical numbers that you have implemented in the package.

File: blog.md (hunk 1 above)
Comment (xuanzhaogao) [reply blog.md]: The beginning is revised, and I added a subsection to introduce Julia programing language.

File: blog.md (hunk 2 above)
Comment (xuanzhaogao) [reply blog.md]: added def of semi ring algebra and tropical algebra, listed the tropical types we implemented

>[Tropical algebra](https://en.wikipedia.org/wiki/Tropical_geometry) is a set of [semiring algebra](https://en.wikipedia.org/wiki/Semiring), which is a generalization of a ring, dropping the requirement that each element must have an additive inverse.
Generally, they can be defined as a set $R$ equipped with two binary operations $\oplus$ and $\otimes$, called addition and multiplication, such that:

* $(R, \oplus)$ is a monoid with identity element called $\mathbb{0}$;
* $(R, \otimes)$ is a monoid with identity element called $\mathbb{1}$;
* Addition is commutative;
* Multiplication by the additive identity $\mathbb{0}$ annihilates ;
* Multiplication left- and right-distributes over addition;
* Explicitly stated, $(R, \oplus)$ is a commutative monoid.


>A Topical algebra can be described as a tuple $(R, \oplus, \otimes, \mathbb{0}, \mathbb{1})$, where $R$ is the set, $\oplus$ and $\otimes$ are the opeartions and $\mathbb{0}$, $\mathbb{1}$ are their identity element, respectively. 
Here some commonly used algebras are listed below:
* `TropicalAndOr`: $([T, F], \lor, \land, F, T)$;
* `TropicalMaxPlus` (also simply called `Tropical`): $(\mathbb{R}, \max, +, -\infty, 0)$;
* `TropicalMinPlus`: $(\mathbb{R}, \min, +, \infty, 0)$;
* `TropicalMaxMul`: $(\mathbb{R}^+, \max, \times, 0, 1)$.

File: blog.md (hunk 3 above)
Comment (xuanzhaogao) [reply blog.md]: added the arXiv link of the ML articles.

File: blog.md (hunk 4 above)
Comment (xuanzhaogao) [reply blog.md]: In the revised version, I mainly focused on 'how to use' instead of 'how to develop'.

File: blog.md (hunk 5 above)
Comment (xuanzhaogao) [reply blog.md]: link added

File: blog.md (hunk 8 above)
Comment (xuanzhaogao) [reply blog.md]: added an example of type system and multiple dispatch
>Type system of `Julia` allows us to create our own types, and `Julia` support multiple dispatch based on that.
For example, we can do the following things in `Julia`
```julia
julia> struct bird end

julia> struct dog end

julia> function fly(a::bird) return true end
fly (generic function with 1 method)

julia> function fly(a::dog) return false end
fly (generic function with 2 methods)

julia> bird_1 = bird(); dog_1 = dog();

julia> fly(bird_1)
true

julia> fly(dog_1)
false
```

File: blog.md (hunk 10 above)
Comment (xuanzhaogao) [reply blog.md]: Detailed data added.
>As we all know, Nvidia GPU is fast because it have a large amount of cuda cores.
For example, the Nvidia A800 GPU have $6912$ cuda cores, which is much more larger than the number of CPU cores, which is normally less than $100$.
The peak flops of A800 is about~$19.5$T, which is also much higher than that of CPU, for example $0.5$T for an Intel i7-10700K.

File: blog.md (hunk 11 above)
Comment (xuanzhaogao) [reply blog.md:160]: Yeah, maybe we can re-export that later.

File: blog.md (hunk 12 above)
Comment (xuanzhaogao) [reply blog.md]: Detailed introduction added.


Another imortant aspect is the memory of GPU, there are three type of programable memories on GPU:
* Global memory, shared by all blocks, large but slow.
* Shared memory, shared by threads in the same block, small but fast, just like a programable L1 cache.
* Registers, used by only one thread.

Driectly loading data from global memory will take a long time, the bandwidth is only about $1000$GB/s and the latency is about $600$ cycles on A800 GPU.
For shared memory, its bandwidth is about $20000$GB/s while the latency is only about $23$ cycles, so that another main target in GPU programing is to properly use the memory, avoid directly loading from the global memory and pre-load the needed data into the shared memory.

File: blog.md (hunk 13 above)
Comment (xuanzhaogao) [reply blog.md]: Explanation added.

GEMM is a very memory-intensive operation, for example, when calculating the GEMM between a $M \times K$ matrix and a $K \times N$ matrix, if we use the naive way, i.e. evaluating the element in the result matrix one by one, for each element we will have to load $2K$ data, and there are $M \times N$ elements in total.

File: blog.md (hunk 15 above)
Comment (xuanzhaogao) [reply blog.md]: I moved that to above, after I finished the discussion about tiling.

File: blog.md (hunk 16 above)
Comment (xuanzhaogao) [reply blog.md]: That's a mistake, I removed that part, and showed an example in the 'how to use' part

File: blog.md [hunk 20]
```diff
@@ -0,0 +1,346 @@
+# How to implement generic matrix multiplication (GEMM) with generic element types on GPU?
+
+After reading this blog, you will learn:
+
+1. What is Tropical algebra? Why we need that?
+2. What is `Julia` programing language and why it is useful?
+3. Why GPU is fast and how to implement a fast generic matrix multiplication on Nvidia GPU?
+4. How to join the [Open Source Promotion Plan](https://summer-ospp.ac.cn/)?
+
+This blog is a technical note for the [Open Source Promotion Plan 2023](https://summer-ospp.ac.cn/) project ["TropicalGEMM on GPU"](https://summer-ospp.ac.cn/org/prodetail/23fec0105?lang=en&list=pro) relesed by JuliaCN, where I developed a [Julia](https://julialang.org/) package [CuTropicalGemm.jl](github.com/TensorBFS/CuTropicalGEMM.jl) calculate Generic Matrix Multiplication (GEMM) of Tropical Numbers on Nvidia GPUs.
```
Comment (GiggleLiu) [inline blog.md]:
```suggestion
This blog is a technical note for the [Open Source Promotion Plan 2023](https://summer-ospp.ac.cn/) project ["TropicalGEMM on GPU"](https://summer-ospp.ac.cn/org/prodetail/23fec0105?lang=en&list=pro) released by JuliaCN, where I developed a [Julia](https://julialang.org/) package [CuTropicalGemm.jl](github.com/TensorBFS/CuTropicalGEMM.jl) calculate Generic Matrix Multiplication (GEMM) of Tropical Numbers on Nvidia GPUs.
```

Comment (GiggleLiu) [reply blog.md]: It looks more natural to move the line 10 to the top.

File: blog.md [hunk 21]
```diff
@@ -0,0 +1,346 @@
+# How to implement generic matrix multiplication (GEMM) with generic element types on GPU?
+
+After reading this blog, you will learn:
```
Comment (GiggleLiu) [inline blog.md]:
```suggestion
This blog covers the following contents
```

File: blog.md [hunk 22]
```diff
@@ -0,0 +1,346 @@
...
+* `TropicalMinPlus`: $(\mathbb{R}, \min, +, \infty, 0)$;
+* `TropicalMaxMul`: $(\mathbb{R}^+, \max, \times, 0, 1)$.
+
+In recent years, the tropical numbers have been widely used in various areas, including optimization, physics, and computer science, due to its computational simplicity.
+
+For example, it was shown that solving the groud state energy of spin glass problem can be mapped as contraction of a tropical tensor network, which is actually calculating tons of matrix multiplication of Tropical numbers, as shown in
+* [arxiv: Tropical Tensor Network for Ground States of Spin Glasses](https://arxiv.org/abs/2008.06888)
+* [Lei Wang: "Tropical Tensor Networks"](https://www.youtube.com/watch?v=l_7xZ4trcnE)
+
+Although such reformulation did not actucally reduce the complexity of the problem, but in this way we can fully use the power of parallel computing technology developed in recent years and greatly speed up the computation.
```
Comment (GiggleLiu) [inline blog.md]:
```suggestion
However, these works are built on top of a poorly implemented tropical matrix multiplication routine that can only use <10% theoreical computational power of a GPU device (check the benchmark bellow). With a proper implement the tropical GEMM on CUDA, these algorithms could be sped up for more than 1 orders.
```

File: blog.md [hunk 23]
```diff
@@ -0,0 +1,346 @@
...
+
+Briefly, `Julia` is fast, and easy to use.
+
+![Julia_Meme](figs/Julia_meme.png)
+
+### User Friendly Tropical Interface: `TropicalNumbers.jl`
+
+For convenience, we want to use tropical number just like normal numbers willout lose of performance, so that we make use of the type system in `Julia`.
+
+Type system of `Julia` allows us to create our own types, and `Julia` support multiple dispatch based on that.
```
Comment (GiggleLiu) [inline blog.md:75]: We do not need this example. BTW: this is am example of single dispatch.

File: blog.md [hunk 24]
```diff
@@ -0,0 +1,346 @@
...
+The plan starts in each summer, and the application commonly opens in June.
+If interested, just go and join OSPP $2024$!
+
+## Acknowledgement
+
+I am very grateful for the guidance and assistance provided by Prof. [Jinguo Liu](https://github.com/GiggleLiu) during the project implementation process.
+I would like to thank [Tim Besard](https://github.com/maleadt) for his invaluable guidance and support during the development of the package, his expertise in GPU utilization have been immensely helpful. 
+I also want to thank [Tyler Thomas](https://github.com/tylerjthomas9) for his assistance in understanding the usage of BinaryBuilder.jl.
+
+This program is support by [Open Source Promotion Plan 2023](https://summer-ospp.ac.cn/), JuliaCN.
```
Comment (GiggleLiu) [inline blog.md]:
```suggestion
```

Repeated information.

File: blog.md (hunk 23 above)
Comment (GiggleLiu) [reply blog.md:75]: Without the example, you could say
```
In Julia, the operators $+$ and $*$ could overloaded elegantly with
[multiple dispatch](link to resource), which a feature not available
in object oriented languages such as Python and C++.
```

**Review by GiggleLiu — APPROVED** [review]
In general, this blog is in good shape, maybe still a bit wordy in multiple places.
I think it is good to go after removing the added "multiple-dispatch" examples.

File: blog.md (hunk 14 above)
Comment (GiggleLiu) [reply blog.md]: `tiling` is still missing. I think this keyword can help readers find more contents related to optimizing matrix multiplication. Try to mention it somewhere?

Comment (GiggleLiu) [reply blog.md]: You should add a sentense like "We use the tiling technique to ..."

File: blog.md [hunk 25]
```diff
@@ -0,0 +1,346 @@
...
+
+As we mentioned above, we choose to use the `Julia` programing language, which is becoming more and more popular in the past few years.
+Why we choose `Julia`?
+Let's just copy something from the blog [*Why We Create Julia?*](https://julialang.org/blog/2012/02/why-we-created-julia/):
+
+>We are greedy: we want more.
+>
+>We want a language that's open source, with a liberal license. We want the speed of C with the dynamism of Ruby. We want a language that's homoiconic, with true macros like Lisp, but with obvious, familiar mathematical notation like Matlab. We want something as usable for general programming as Python, as easy for statistics as R, as natural for string processing as Perl, as powerful for linear algebra as Matlab, as good at gluing programs together as the shell. Something that is dirt simple to learn, yet keeps the most serious hackers happy. We want it interactive and we want it compiled.
+
+Briefly, `Julia` is fast, and easy to use.
```
Comment (GiggleLiu) [inline blog.md:67]:
```suggestion
Briefly, `Julia` is fast, and easy to use.
```
The paragraph is sharp enough because it is disconnected with the rest.

The best way is to copy the introduction to Julia in a paper (e.g. [the Yao paper](https://arxiv.org/pdf/1912.10877.pdf)):
image

Another very good example is [CUDA.jl paper](https://arxiv.org/pdf/1712.03112.pdf):
image

**Comment by xuanzhaogao** [comment]
Thanks a lot.
The mentioned part are revised and the example of multiple dispatch is removed.
